src/crimson/os/seastore/cache.h

```diff
@@ -181,42 +181,47 @@ class Cache {
     paddr_t offset,       ///< [in] starting addr
     segment_off_t length  ///< [in] length
   ) {
-    if (auto iter = extents.find_offset(offset);
-	       iter != extents.end()) {
-      auto ret = TCachedExtentRef<T>(static_cast<T*>(&*iter));
+    auto cached = query_cache_with_placeholders(offset);
+    if (!cached) {
+      auto ret = CachedExtent::make_cached_extent_ref<T>(
+        alloc_cache_buf(length));
+      ret->set_paddr(offset);
+      ret->state = CachedExtent::extent_state_t::CLEAN;
+      add_extent(ret);
+      return read_extent<T>(std::move(ret));
+    }
+
+    // extent PRESENT in cache
+    if (cached->get_type() == extent_types_t::RETIRED_PLACEHOLDER) {
+      auto ret = CachedExtent::make_cached_extent_ref<T>(
+        alloc_cache_buf(length));
+      ret->set_paddr(offset);
+      ret->state = CachedExtent::extent_state_t::CLEAN;
+      extents.replace(*ret, *cached);
+
+      // replace placeholder in transactions
+      std::vector<Transaction*> cached_trans(cached->transactions.size());
+      std::transform(cached->transactions.begin(),
+                     cached->transactions.end(),
+                     cached_trans.begin(),
+                     [] (const read_set_item_t<Transaction>& item) {
+                       return item.t;
+                     });
+      for (auto& t: cached_trans) {
+        t->replace_placeholder(*cached, *ret);
+      }
+      assert(cached->transactions.empty());
+      cached->state = CachedExtent::extent_state_t::INVALID;
+      return read_extent<T>(std::move(ret));
+    } else {
+      auto ret = TCachedExtentRef<T>(static_cast<T*>(cached.get()));
       return ret->wait_io(
-      ).then([ret=std::move(ret)]() mutable -> get_extent_ret<T> {
-	// ret may be invalid, caller must check
-	return get_extent_ret<T>(
-	  get_extent_ertr::ready_future_marker{},
-	  std::move(ret));
+      ).then([ret=std::move(ret)] () mutable -> get_extent_ret<T> {
```

Comment:
No space after ], check the rest of the series.

Reply:
fixed